### ISSUE TITLE
Writing "&" character in entry title fails

### ISSUE THREAD

**Issue by joonas-fi** [tobischo/gokeepasslib#22]
Fails = Keepass tells that the file is corrupted.

Probably also fails with any other elements (UserName, Password), in `Entry`'s `Values` member.

Code to reproduce: https://github.com/tobischo/gokeepasslib/blob/master/examples/example-writing.go (change "My GMail password" to "Foo & Bar").

`Foo &amp; Bar` works, so apparently the user of the API has to be aware of the XML-backedness, which was a bit surprising to me.

Shouldn't gokeepasslib do the XML-escaping? Or if not, maybe we should change `mkValue()` and `mkProtectedValue()` in the file writing example to do the XML-escaping, so not more people stumble with this.

Is this true in the reading direction as well, that when you read the file with `Title` containing `&amp;`, we get that back instead of `&`?

**Comment by joonas-fi** [comment]
Making these changes to the example fixes the issue:

```
// add imports 	"bytes" and "encoding/xml"

func xmlEscape(value string) string {
	escaped := &bytes.Buffer{}

	if err := xml.EscapeText(escaped, []byte(value)); err != nil {
		panic(err)
	}

	return escaped.String()
}

func mkValue(key string, value string) gokeepasslib.ValueData {
	return gokeepasslib.ValueData{Key: key, Value: gokeepasslib.V{Content: xmlEscape(value)}}
}

func mkProtectedValue(key string, value string) gokeepasslib.ValueData {
	return gokeepasslib.ValueData{Key: key, Value: gokeepasslib.V{Content: xmlEscape(value), Protected: true}}
}
```

But I would think this is supposed to be implemented at lower level than the API level? I.e. I didn't think gokeepasslib.ValueData is raw XML.

**Comment by tobischo** [comment]
Oh, wow, excellent catch.

That might actually have been the one issue which always broke my personal keepass file when trying to write it with gokeepasslib using [kp2](https://github.com/tobischo/kp2) (and also the reason why I haven't added file writing there)

You are right, it does seem counter intuitive to have to know that the data is stored in XML.
We should take a look at where keepass2 is doing this

**Comment by tobischo** [comment]
Can't find anything specific about it at first glance.
Mostly found references to such replacements in data exchange parts.

I'll take a look this evening when I have more time - hopefully this evening

**Comment by tobischo** [comment]
OK, it is definitely unrelated to the encryption issue I experienced.
You opening this issue up caused me to look into that as well though! Found that bug at least 😄 Thank you!

**Comment by joonas-fi** [comment]
Glad to know that I was of any help :)

At least here's how keepassx does it:

[KeePass2Writer::writeDatabase()](https://github.com/keepassx/keepassx/blob/b654fde9cd43202903c0e9e08b2043315335759b/src/format/KeePass2Writer.cpp#L43)

which calls [KeePass2XmlWriter::writeDatabase()](https://github.com/keepassx/keepassx/blob/b654fde9cd43202903c0e9e08b2043315335759b/src/format/KeePass2XmlWriter.cpp#L38)

which calls writeGroup()

which calls [writeEntry()](https://github.com/keepassx/keepassx/blob/b654fde9cd43202903c0e9e08b2043315335759b/src/format/KeePass2XmlWriter.cpp#L329)

which calls [QXmlStreamWriter::writeCharacters()](http://doc.qt.io/qt-4.8/qxmlstreamwriter.html#writeCharacters)

which is documented as:

> Writes text. The characters "<", "&", and """ are escaped as entity references "&lt;", "&amp;, and "&quot;". To avoid the forbidden sequence "]]>", ">" is also escaped as "&gt;".

**Comment by joonas-fi** [comment]
Golang's XML marshalling by default definitely escapes any characters that need to be escaped for XML. So I noticed that in https://github.com/tobischo/gokeepasslib/blob/master/content.go#L279 there's "innerxml" defined, which suppresses that default behaviour. I presume that it's for the "encryption inside already encrypted container" -feature.

**Comment by tobischo** [comment]
`innerxml` is used to be able to add content inside the tag in addition to attributes.

As seen here:
https://github.com/tobischo/gokeepasslib/blob/master/examples/example.xml#L97

Having read through the descriptions at 
* https://golang.org/pkg/encoding/xml/#Unmarshal
* https://golang.org/pkg/encoding/xml/#Marshal
`chardata` seems to be the correct choice here.

Example: https://play.golang.org/p/-j3xNDd8Hb

What do you think?

**Comment by joonas-fi** [comment]
Oh yes ok now I understand better how XML marshaling works in golang. So basically there can only be one `innerxml` or `chardata` struct tag within a single struct, and that one controls whether that field is an attribute or the tag's content. Yeah, and I can see it's pretty easy to get mixed up if trying `innerxml` first and it seems to work (until one needs special characters :) ).

Yeah, so `innerxml` -> `chardata` would be the fix? If so, LGTM.

**Comment by tobischo** [comment]
At least that was my hope, but using `chardata` also encodes e.g. newlines to `&#xA;` breaking the file when trying to open it with Keepass2.

I believe the solution to go for is `innerxml` and a custom type for unmarshaling and marshaling

**Comment by tobischo** [comment]
My current idea is a custom (un-)marshaler on `V`

```go
// V is a wrapper for the content of a value, so that it can store whether it is protected
type V struct {
	Content   string      `xml:",innerxml"`
	Protected boolWrapper `xml:"Protected,attr,omitempty"`
}

type vContentMapping struct {
	Decoded string
	Encoded string
}

var vContentMappings = []vContentMapping{
	{Decoded: `&`, Encoded: `&amp;`},
	{Decoded: `<`, Encoded: `&lt;`},
	{Decoded: `>`, Encoded: `&gt;`},
	{Decoded: `"`, Encoded: `&quot;`},
}

func (v *V) UnmarshalXML(d *xml.Decoder, start xml.StartElement) error {
	for _, attr := range start.Attr {
		switch attr.Name.Local {
		// There shouldn't be other cases but this feels more readable than an if statement...
		case "Protected":
			err := v.Protected.UnmarshalXMLAttr(attr)
			if err != nil {
				return err
			}
		}
	}

	var content string
	err := d.DecodeElement(&content, &start)
	if err != nil {
		return err
	}

	for _, mapping := range vContentMappings {
		content = strings.Replace(content, mapping.Encoded, mapping.Decoded, -1)
	}

	v.Content = content

	return nil
}

func (v *V) MarshalXML(e *xml.Encoder, start xml.StartElement) error {
	protectedAttr, err := v.Protected.MarshalXMLAttr(xml.Name{Local: "Protected"})
	if err != nil {
		return err
	}
	start.Attr = []xml.Attr{protectedAttr}

	content := v.Content
	for _, mapping := range vContentMappings {
		content = strings.Replace(content, mapping.Decoded, mapping.Encoded, -1)
	}

	return e.EncodeElement(content, start)
}
```
This doesn't solve the issue though.
I can open the file just right, but the values of the passwords are broken when opening the file again using Keepass2 after unmarshaling, unlocking, locking and remarshaling.

Interestingly when opening the same file with gokeepasslib, I get the password out just right

**Comment by joonas-fi** [comment]
I don't quite understand how the passwords can be broken. The passwords shouldn't ever be affected (assuming you store passwords as Protected:true), because they're represented in the XML as base64, and its alphabet (`A-Za-z0-9+/=`) doesn't contain any escapeable characters.

**Comment by joonas-fi** [comment]
I'm sorry if I'm not being much of a help, but I'm caught with this:

> At least that was my hope, but using chardata also encodes e.g. newlines to &#xA; breaking the file when trying to open it with Keepass2.

This already sounds fishy to me, because XML is XML, and decoding that should yield \n in the program that reads it (and since Keepass2 is C# I'd assume the XML parser works correctly because Microsoft relies so much on XML).

AFAIK \n needs to be encoded because reading `<tag>\nfoo</tag>` would yield `foo` (= newline stripped) because leading and trailing whitespace is supposed to be ignored, therefore it's just easier to encode strings like `foo\nbar` to `foo&#xA;bar` as well (even though mid-whitespace doesn't require it). That behaviour does sound legitimate to me, and therefore I can't stop to think that `chardata` tag should be enough and maybe the issue is somewhere deeper?

**Comment by tobischo** [comment]
Regarding the broken passwords: I am confident that that is unrelated to the XML issue.
Having looked at the decrypted XML and the XML being encoded, the content of the passwords is identical and doesn't have to concern us here any further.

**Comment by tobischo** [comment]
I believe I have found the issues...

1. `chardata` seems to work and the issues I experienced seem to have been unrelated.
2. `AutoTypeAssociation` apparently should be optional/omitempty (not necessary but helps when comparing files)
3. `Entries` should come before `Groups` as nested elements for `Group`: after changing that the broken password issues I experienced were resolved - this also better matches the XML format Keepass2 creates - although I do not really understand why that would be an issue at the moment
